WiiU/WiiUHost.cpp

```diff
@@ -0,0 +1,215 @@
+		NKCODE_BUTTON_SELECT, // VPAD_BUTTON_MINUS_BIT            = 2,
+		NKCODE_BUTTON_START,  // VPAD_BUTTON_PLUS_BIT             = 3,
+		NKCODE_BUTTON_R1,     // VPAD_BUTTON_R_BIT                = 4,
+		NKCODE_BUTTON_L1,     // VPAD_BUTTON_L_BIT                = 5,
+		NKCODE_BUTTON_R2,     // VPAD_BUTTON_ZR_BIT               = 6,
+		NKCODE_BUTTON_L2,     // VPAD_BUTTON_ZL_BIT               = 7,
+		NKCODE_DPAD_DOWN,     // VPAD_BUTTON_DOWN_BIT             = 8,
+		NKCODE_DPAD_UP,       // VPAD_BUTTON_UP_BIT               = 9,
+		NKCODE_DPAD_RIGHT,    // VPAD_BUTTON_RIGHT_BIT            = 10,
+		NKCODE_DPAD_LEFT,     // VPAD_BUTTON_LEFT_BIT             = 11,
+		NKCODE_BUTTON_Y,      // VPAD_BUTTON_Y_BIT                = 12,
+		NKCODE_BUTTON_X,      // VPAD_BUTTON_X_BIT                = 13,
+		NKCODE_BUTTON_B,      // VPAD_BUTTON_B_BIT                = 14,
+		NKCODE_BUTTON_A,      // VPAD_BUTTON_A_BIT                = 15,
+		NKCODE_UNKNOWN,       // VPAD_BUTTON_TV_BIT               = 16,
+		NKCODE_BUTTON_THUMBR, // VPAD_BUTTON_STICK_R_BIT          = 17,
+		NKCODE_BUTTON_THUMBL, // VPAD_BUTTON_STICK_L_BIT          = 18,
+		NKCODE_UNKNOWN,       // VPAD_BUTTON_TOUCH_BIT            = 19,
+		NKCODE_UNKNOWN,       // VPAD_BUTTON_UNUSED1_BIT          = 20,
+		NKCODE_UNKNOWN,       // VPAD_BUTTON_UNUSED2_BIT          = 21,
+		NKCODE_UNKNOWN,       // VPAD_BUTTON_UNUSED3_BIT          = 22,
+		NKCODE_UNKNOWN,       // VPAD_STICK_R_EMULATION_DOWN_BIT  = 23,
+		NKCODE_UNKNOWN,       // VPAD_STICK_R_EMULATION_UP_BIT    = 24,
+		NKCODE_UNKNOWN,       // VPAD_STICK_R_EMULATION_RIGHT_BIT = 25,
+		NKCODE_UNKNOWN,       // VPAD_STICK_R_EMULATION_LEFT_BIT  = 26,
+		NKCODE_UNKNOWN,       // VPAD_STICK_L_EMULATION_DOWN_BIT  = 27,
+		NKCODE_UNKNOWN,       // VPAD_STICK_L_EMULATION_UP_BIT    = 28,
+		NKCODE_UNKNOWN,       // VPAD_STICK_L_EMULATION_RIGHT_BIT = 29,
+		NKCODE_UNKNOWN,       // VPAD_STICK_L_EMULATION_LEFT_BIT  = 30,
+	};
+
+	VPADStatus vpad;
+	VPADReadError readError;
+	VPADRead(chan, &vpad, 1, &readError);
+
+	if (!readError) {
+		static int touchflags;
+		if (vpad.tpFiltered1.validity != VPAD_VALID) {
+			vpad.tpFiltered1.touched = false;
+		}
+		if (touchflags == TOUCH_DOWN || touchflags == TOUCH_MOVE) {
+			touchflags = vpad.tpFiltered1.touched ? TOUCH_MOVE : TOUCH_UP;
+		} else {
+			touchflags = vpad.tpFiltered1.touched ? TOUCH_DOWN : 0;
+		}
+		if (touchflags) {
+			VPADTouchData calibrated;
+			VPADGetTPCalibratedPointEx(chan, VPAD_TOUCH_RESOLUTION_854x480, &calibrated, &vpad.tpFiltered1);
+			NativeTouch({ (float)calibrated.x, (float)calibrated.y, 0, touchflags });
+		}
+		for (int i = 0; i < countof(keymap); i++) {
+			if (keymap[i] == NKCODE_UNKNOWN)
+				continue;
+
+			if ((vpad.trigger | vpad.release) & (1 << i)) {
+				NativeKey({ DEVICE_ID_PAD_0 + chan, keymap[i], (vpad.trigger & (1 << i)) ? KEY_DOWN : KEY_UP });
+			}
+		}
+		NativeAxis({ DEVICE_ID_PAD_0 + chan, JOYSTICK_AXIS_X, vpad.leftStick.x});
```

Comment:
If possible, send these ONLY when they change.

The reason is, let's assume I map another button to analog, or let's say I use touch screen controls. These will override whatever came from touch or the other button because you're actively saying the analog stick is being moved.

-[Unknown]

Reply:
oh okay. was just being a bit lazy there, didn't think it would matter much xD